Refactor drain on shutdown

wjordan opened the pull request:
### Description
This refactoring PR moves the `drain_on_shutdown` implementation out of `#graceful_shutdown` to reduce duplication with the main server accept loop, including a test covering the basic behavior of this option to prevent regressions.

There might be some slight change in timing around when exactly the `:state` event gets fired (before or after connections are all drained), but this doesn't seem important enough to worry about (there was zero test coverage around it).

- dad2708 refactors `TestPumaServer#server_run`, in order to allow arbitrary option arguments to be passed. This made writing the added test with a custom option a bit easier and more readable, and allowed me to clean up a bit of boilerplate in the rest of the test suite.

### Your checklist for this pull request
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

MSP-Greg commented on test/test_puma_server.rb:
```diff
@@ -146,7 +145,7 @@ def test_default_server_port
 
   def test_default_server_port_respects_x_forwarded_proto
-    server_run app: ->(env) do
+    server_run do |env|
```
Missed one (I've done similar)? Maybe:
```ruby
server_run { [200, {}, [env['SERVER_PORT']]] }
```

MSP-Greg reviewed (commented):
Any idea about the Ruby 2.2 fails?  Otherwise, LGTM.

MSP-Greg commented:
@wjordan

Maybe change the test to the following?  Passed in my fork...

```ruby
def test_drain_on_shutdown
  server_run(drain_on_shutdown: true, max_threads: 1) do
    sleep 0.005
    [200, {}, ["DONE"]]
  end
  connections = Array.new(20) { send_http "GET / HTTP/1.0\r\n\r\n" }
  @server.stop
  @thread.join
  connections.each do |s|
    assert_match 'DONE', s.read
  end
end

def test_not_drain_on_shutdown
  server_run(max_threads: 1) do
    sleep 0.005
    [200, {}, ["DONE"]]
  end
  connections = Array.new(20) { send_http "GET / HTTP/1.0\r\n\r\n" }
  @server.stop
  @thread.join
  bad = 0
  connections.each do |s|
    begin
      assert_match 'DONE', s.read
    rescue Errno::ECONNRESET
      bad += 1
    end
  end
  assert_operator bad, :>, 5
end
```

wjordan commented:
> Any idea about the Ruby 2.2 fails?

Oops! `Queue#close` was added in Ruby 2.3, I had earlier written a simple backport patch (`Puma::QueueClose`) but the simple patch did not un-block already-waiting threads when close was called as expected. I added a few extra lines to the patch which fixes the test on Ruby 2.2. Thanks for taking a look!

I also added a `test_not_drain_on_shutdown` based on your patch, thanks!

MSP-Greg commented:
Please have a look at the errors in the 2.2 test logs.  I added the `sleep 0.005` as it only adds 100 mS to each test, and pretty much forces the behavior we want, regardless of issues with test runners, etc, etc...  Or, I think the `test_drain_on_shutdown` test in the patch should work in CI.

Something is amiss with my Windows Ruby 2.2 (build system) on the new desktop, I can't build Puma, so I just used CI...

wjordan commented:
> Please have a look at the errors in the 2.2 test logs.

Oops again- I messed up the `QueueClose#pop` implementation (returns `nil` even if there are items in the closed queue), fixed. (Hope Ruby 2.2 support can be dropped soon...)

While looking at the error log for this, I may have spotted a separate race condition in `Reactor#select_loop`- I'll work on that issue separately, but I'm glad this detour led to that discovery!
